File: Documentation/3.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Querying/HowToQuery.cs
using System.Linq;

using Raven.Client.Document;
using Raven.Client.Indexes;
using Raven.Client.Linq;
using Raven.Documentation.CodeSamples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Session.Querying
{
	public class HowToQuery
	{
		private class MyCustomIndex : AbstractIndexCreationTask<Employee>
		{
		}

		private interface IFoo
		{
			#region query_1_0
			IRavenQueryable<T> Query<T>(); // perform query on a dynamic index

			IRavenQueryable<T> Query<T>(string indexName, bool isMapReduce = false);

			IRavenQueryable<T> Query<T, TIndexCreator>()
				where TIndexCreator : AbstractIndexCreationTask, new();
			#endregion
		}

		public HowToQuery()
		{
			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region query_1_1
					// load up to 128 entities from 'Employees' collection
					var employees = session
						.Query<Employee>()
						.ToList();
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region query_1_2
					// load up to 128 entities from 'Employees' collection
					// where FirstName equals 'Robert'
					var employees = session
						.Query<Employee>()
						.Where(x => x.FirstName == "Robert")
						.ToList();
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region query_1_3
					// load up to 128 entities from 'Employees' collection
					// where FirstName equals 'Robert'
					var employees = from employee in session.Query<Employee>()
								 where employee.FirstName == "Robert"
								 select employee;
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region query_1_4
					// load up to 128 entities from 'Employees' collection
					// where FirstName equals 'Robert'
					// using 'My/Custom/Index'
					var employees = from employee in session.Query<Employee>("My/Custom/Index")
								 where employee.FirstName == "Robert"
								 select employee;
					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region query_1_5
					// load up to 128 entities from 'Employees' collection
					// where FirstName equals 'Robert'
					// using 'My/Custom/Index'
					var employees = from employee in session.Query<Employee, MyCustomIndex>()
								 where employee.FirstName == "Robert"
								 select employee;
					#endregion
				}
			}
		}
	}
}

File: Documentation/3.0/Samples/java/src/test/java/net/ravendb/clientapi/session/querying/HowToQuery.java
package net.ravendb.clientapi.session.querying;

import java.util.List;

import net.ravendb.client.IDocumentSession;
import net.ravendb.client.IDocumentStore;
import net.ravendb.client.document.DocumentStore;
import net.ravendb.client.indexes.AbstractIndexCreationTask;
import net.ravendb.client.linq.IRavenQueryable;
import net.ravendb.samples.northwind.Employee;
import net.ravendb.samples.northwind.QEmployee;


public class HowToQuery {

  private class MyCustomIndex extends AbstractIndexCreationTask {
    // ...
  }

  @SuppressWarnings("unused")
  private interface IFoo {
    //region query_1_0
    public <T> IRavenQueryable<T> query(Class<T> clazz);  // perform query on a dynamic index

    public <T> IRavenQueryable<T> query(Class<T> clazz, String indexName);

    public <T> IRavenQueryable<T> query(Class<T> clazz, String indexName, boolean isMapReduce);

    public <T> IRavenQueryable<T> query(Class<T> clazz, Class<? extends AbstractIndexCreationTask> indexCreator);
    //endregion
  }

  @SuppressWarnings("unused")
  public HowToQuery() throws Exception {
    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region query_1_1
        // load up to 128 entities from 'Employees' collection
        List<Employee> employees = session
          .query(Employee.class)
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region query_1_2
        // load up to 128 entities from 'Employees' collection
        // where FirstName equals 'Robert'
        QEmployee e = QEmployee.employee;
        List<Employee> employees = session
          .query(Employee.class)
          .where(e.firstName.eq("Robert"))
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region query_1_4
        // load up to 128 entities from 'Employees' collection
        // where FirstName equals 'Robert'
        // using 'My/Custom/Index'
        QEmployee e = QEmployee.employee;
        List<Employee> employees = session
          .query(Employee.class, "My/Custom/Index")
          .where(e.firstName.eq("Robert"))
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region query_1_5
        // load up to 128 entities from 'Employees' collection
        // where FirstName equals 'Robert'
        // using 'My/Custom/Index'
        QEmployee e = QEmployee.employee;
        List<Employee> employees = session
          .query(Employee.class, MyCustomIndex.class)
          .where(e.firstName.eq("Robert"))
          .toList();
        //endregion
      }
    }
  }
}
